### PR TITLE
#166789876  Create an endpoint to allow a user to rate an article

### PR DESCRIPTION
**Pull request by Dom58**
#### What does this PR do?
This pull request helps an authorized user to give a rate to an article: 
 - authorized user should be able to rate an article.
 - the author cannot rate his/her article

#### Description of Task to be completed?
Authorized user should be able to rate once an article .
> the feature can be accessed and tested via [Heroku](https://ah-lobos-backend-swagger.herokuapp.com)

#### How should this be manually tested?
Follow instructions on the `README.md` file attached to this repo to set up the project on your local machine.
Access the following routes to test
> User rating an article `api/articles/<slug>/ratings`

#### Any background context you want to provide?
Author can't rate his article and user must be authorized yet to access this task

#### What are the relevant pivotal tracker stories?
[#166789876](https://www.pivotaltracker.com/story/show/166789876)

#### Screenshots (if appropriate)
* Swagger documentation

Screen Shot 2019-07-14 at 18 04 23

* Rating article using Postman 

Screen Shot 2019-07-14 at 16 27 07

* Mockups
** User should not rate his/her article
Screen Shot 2019-07-15 at 17 39 52

### CODE REVIEW

File: src/controllers/articleControllers.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    const findIfUserIfRatedArticle = await Rating.findOne({
+      where: {
+        user: findUser.id,
+        post: findArticle.id
+      }
+    });
+    if (findIfUserIfRatedArticle) {
+      return res.status(400).json({
+        error: 'You are already rate this article'
```
Comment (j4l13n) [inline src/controllers/articleControllers.js]: - can you change the message of this response?

File: src/db/migrations/20190710174425-create-rating.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        allowNull: false,
+        type: Sequelize.DATE,
+        defaultValue: Sequelize.fn('now')
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('Ratings');
+  }
+};
```
Comment (j4l13n) [inline src/db/migrations/20190710174425-create-rating.js]: - add a new line

**Comment by Emile-Nsengimana** [comment]
is the screen shoot of tests necessary in the PR when we have Travis CI checks?

**Comment by j4l13n** [comment]
> is the screen shoot of tests necessary in the PR when we have Travis CI checks?

**Comment by j4l13n** [comment]
> is the screen shoot of tests necessary in the PR when we have Travis CI checks?

I think, no need for that. also, remember to add your mockups of your task

**Comment by Emile-Nsengimana** [comment]
Is there any particular reason why this PR was closed?

File: src/controllers/articleControllers.js (hunk 1 above)
Comment (Dom58) [reply src/controllers/articleControllers.js]: Yes I can. Review again was changed

File: src/db/migrations/20190710174425-create-rating.js (hunk 2 above)
Comment (Dom58) [reply src/db/migrations/20190710174425-create-rating.js]: > * add a new line

Added

**Review by j4l13n — APPROVED** [review]

File: src/db/models/blacklisttokens.js [hunk 3]
```diff
@@ -8,4 +8,4 @@ module.exports = (sequelize, DataTypes) => {
     // associations can be defined here
   };
   return BlacklistTokens;
-};
```
Comment (adafia) [inline src/db/models/blacklisttokens.js:11]: Please attend to this

**Comment by emukungu** [comment]
what's the purpose of adding a comment when rating, it's not necessary.

**Comment by Dom58** [comment]
> what's the purpose of adding a comment when rating, it's not necessary.



> what's the purpose of adding a comment when rating, it's not necessary.

Thank you for your feedback, I'm going to remove all unnecessary comments

**Review by Emile-Nsengimana — APPROVED** [review]

**Comment by Dom58** [comment]
> > what's the purpose of adding a comment when rating, it's not necessary.
 
> Thank you for your feedback, I'm going to remove all unnecessary comments

**Review by manzif — APPROVED** [review]

File: src/db/models/draftarticles.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+  const Draftarticles = sequelize.define('Draftarticles', {
```
Comment (emukungu) [inline src/db/models/draftarticles.js]: what's the purpose of this table, yet an article's table was created from the 'create articles' task.
Please use the same table, don't create another one.

File: src/db/models/rating.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      type: DataTypes.UUID,
+      allowNull: false,
+      primaryKey: true,
+      defaultValue: DataTypes.UUIDV4
+    },
+    rating: {
+      type: DataTypes.INTEGER,
+      allowNull: false
+    },
+    comment: {
```
Comment (emukungu) [inline src/db/models/rating.js]: Don't think we need to add comments while rating

File: src/middlewares/articleValidation.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    * @returns {Object} ratingSchema object
+    */
+  static async rating(req, res, next) {
+    const ratingSchema = Joi.object().keys({
+      rating: Joi.number().integer().valid('1', '2', '3', '4', '5').label('Rate')
+        .required(),
+      comment: Joi.string().allow('').allow(null)
+    });
+    const {
+      rating, comment
```
Comment (emukungu) [inline src/middlewares/articleValidation.js]: only rate the article, without the comment

File: src/db/models/draftarticles.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      allowNull: false
+    },
+    slug: {
+      type: DataTypes.STRING,
+      allowNull: false
+    }
+  }, {});
+  Draftarticles.associate = function(models) {
+    Draftarticles.hasMany(models.Rating, {
+      foreignKey: 'post'
```
Comment (emukungu) [inline src/db/models/draftarticles.js]: use a more descriptive variable not 'post' for anything attached to the post variable

**Review by emukungu — CHANGES_REQUESTED** [review]

**Comment by emukungu** [comment]
correct the grammar for the PR's heading.

File: src/db/models/draftarticles.js (hunk 4 above)
Comment (Dom58) [reply src/db/models/draftarticles.js]: I used before rebasing develop, let's delete it

File: src/controllers/articleRatingControllers.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      const saveRating = await Rating.create({
+        rating,
+        user: findUser.id,
+        articleId: findArticle.id
+      });
+      return res.status(201).json({
+        message: 'Article rated successfuly',
+        data: {
+          user: saveRating.user,
+          rating: saveRating.rating
```
Comment (emukungu) [inline src/controllers/articleRatingControllers.js]: destructure your code

File: src/controllers/articleRatingControllers.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    const { slug } = req.params;
+    const { rating } = req.body;
+
+    const findArticle = await Articles.findOne({ where: { slug } });
+    if (!findArticle) {
+      return res.status(404).send({
+        error: 'Article not found!'
+      });
+    }
+    const findUser = await Users.findOne({ where: { email: req.user.email } });
```
Comment (emukungu) [inline src/controllers/articleRatingControllers.js]: destructure your code

File: src/controllers/articleRatingControllers.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import model from '../db/models/index';
+
+const { Users, Articles, Rating } = model;
+
+/**
+ * rating controller
+ */
+class ArticleManager {
```
Comment (emukungu) [inline src/controllers/articleRatingControllers.js]: change the name to reflect your task.

File: src/middlewares/articleValidation.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import Joi from '@hapi/joi';
+
+/**
+ *  rating article validaton
+ */
+class ArticleValidation {
```
Comment (emukungu) [inline src/middlewares/articleValidation.js]: change the class name to reflect your task

File: src/controllers/articleRatingControllers.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        articleId: findArticle.id
+      }
+    });
+    if (findIfUserRatedArticle) {
+      return res.status(400).json({
+        error: 'You have already rated this article'
+      });
+    }
+
+    const articleRatings = await Rating.findOne({ where: { articleId: findArticle.id, rating } });
```
Comment (emukungu) [inline src/controllers/articleRatingControllers.js]: what's the purpose of this line?

**Review by emukungu — CHANGES_REQUESTED** [review]

File: src/controllers/articleRatingControllers.js (hunk 12 above)
Comment (Dom58) [reply src/controllers/articleRatingControllers.js]: This line helps in finding a user rating an article before, with the same rate value requested

Comment (emukungu) [reply src/controllers/articleRatingControllers.js]: Don't think this check is necessary, since you have already blocked the user from rating again.

**Review by emukungu — CHANGES_REQUESTED** [review]

**Pull request merged**